builder/bundleset.go

```diff
@@ -114,46 +125,150 @@ func sortBundles(bundles bundleSet) ([]*bundle, error) {
+	}
+
+	return nil
+}
+
+func validateBundleName(name string) error {
+	if !validBundleNameRegex.MatchString(name) || name == "MoM" || name == "full" {
+		return fmt.Errorf("Invalid bundle name %q", name)
+	}
+
+	return nil
+}
+
+func validateBundle(b *bundle, strictLvl ValidationLevel) error {
+	var errText string
+
+	// Low level strictness
+	if err := validateBundleName(b.Header.Title); err != nil {
+		errText = err.Error() + " in bundle header Title"
+	}
+	if strictLvl == StrictnessLow {
+		if errText != "" {
+			return errors.New(strings.TrimSuffix(errText, "\n"))
+		}
+		return nil
+	}
+
+	// Medium level strictness
+	if b.Header.Description == "" {
+		errText += "Empty Description in bundle header\n"
+	}
+	if b.Header.Maintainer == "" {
+		errText += "Empty Maintainer in bundle header\n"
+	}
+	if strictLvl == StrictnessMed {
+		if errText != "" {
+			return errors.New(strings.TrimSuffix(errText, "\n"))
+		}
+		return nil
+	}
+
+	// High level strictness
+	if b.Header.Status == "" {
+		errText += "Empty Status in bundle header\n"
+	}
+	if b.Header.Capabilities == "" {
+		errText += "Empty Capabilities in bundle header\n"
+	}
+	if errText != "" {
+		return errors.New(strings.TrimSuffix(errText, "\n"))
+	}
+
+	return nil
+}
+
 // parseBundleFile parses a bundle file identified by full filepath, and returns
 // a bundle object representation of that bundle.
 // Note: the Allpackages field of the bundles are left blank, as they cannot
 // be calculated in isolation.
 func parseBundleFile(filename string) (*bundle, error) {
 	name := filepath.Base(filename)
-	if !validBundleNameRegex.MatchString(name) || name == "MoM" || name == "full" {
-		return nil, fmt.Errorf("invalid bundle name %q derived from file %s", name, filename)
-	}
 
 	contents, err := ioutil.ReadFile(filename)
 	if err != nil {
 		return nil, err
 	}
 
-	includes, packages, err := parseBundle(contents)
+	bundle, err := parseBundle(contents)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't parse bundle file %s: %s", filename, err)
+		return nil, fmt.Errorf("Couldn't parse bundle file %s: %s", filename, err)
 	}
 
-	b := &bundle{
-		Name:           name,
-		DirectIncludes: includes,
-		DirectPackages: packages,
-		Filename:       filename,
-	}
-	return b, nil
+	bundle.Name = name
+	bundle.Filename = filename
+
+	return bundle, nil
 }
 
 // parseBundle parses the bytes of a bundle file, ignoring comments and
 // processing "include()" directives the same way that m4 works.
-func parseBundle(contents []byte) (includes []string, packages []string, err error) {
+func parseBundle(contents []byte) (*bundle, error) {
 	scanner := bufio.NewScanner(bytes.NewReader(contents))
 
+	var b bundle
+	var includes, packages []string
+
 	line := 0
 	for scanner.Scan() {
 		line++
 		text := scanner.Text()
 		comment := strings.Index(text, "#")
 		if comment > -1 {
+			if matches := bundleHeaderFieldRegex.FindStringSubmatch(text); len(matches) > 2 {
```

Comment:
Optional: I'd also only match the comment properties if the line starts with a comment. Otherwise you need think about lines like:  `package   # TITLE: name`.

Reply:
The regex is `^# \[([A-Z]+)\]:\s*(.*)$`, so it does only match if it starts with a comment. One of my tests uses
```
# [TITLE]: realtitle
# [DESCRIPTION]: a description
# [STATUS]: 
# [CAPABILITIES]: 
# [MAINTAINER]: 
include(a)
pkg1 # [TITLE]: wrongtitle
```
To make sure that lines like that aren't checked.

Reply:
Ah, haven't noticed the `^`. I'd still add the check, but indeed it is not important for correctness.